sql/2024/04/inaccurate-sizes-attribute-impact.sql

```diff
@@ -0,0 +1,106 @@
+# limitations under the License.
+
+
+CREATE TEMPORARY FUNCTION
+  getSrcsetSizesAccuracy(payload STRING)
+  RETURNS ARRAY<STRUCT<sizesAbsoluteError INT64,
+  sizesRelativeError FLOAT64,
+  wDescriptorAbsoluteError INT64,
+  wDescriptorRelativeError FLOAT64,
+  actualSizesEstimatedWastedLoadedPixels INT64,
+  actualSizesEstimatedWastedLoadedBytes FLOAT64,
+  wastedLoadedPercent FLOAT64>>
+  LANGUAGE js AS '''
+try {
+  var $ = JSON.parse(payload);
+  var responsiveImages = JSON.parse($._responsive_images);
+  responsiveImages = responsiveImages['responsive-images'];
+  return responsiveImages.map(({
+    sizesAbsoluteError,
+    sizesRelativeError,
+    wDescriptorAbsoluteError,
+    wDescriptorRelativeError,
+    idealSizesSelectedResourceEstimatedPixels,
+    actualSizesEstimatedWastedLoadedPixels,
+    actualSizesEstimatedWastedLoadedBytes
+  }) => {
+    let wastedLoadedPercent;
+    if ( idealSizesSelectedResourceEstimatedPixels > 0 ) {
+      wastedLoadedPercent = actualSizesEstimatedWastedLoadedPixels / idealSizesSelectedResourceEstimatedPixels;
+    } else {
+      wastedLoadedPercent = null;
+    }
+    return {
+      sizesAbsoluteError,
+      sizesRelativeError,
+      wDescriptorAbsoluteError,
+      wDescriptorRelativeError,
+      actualSizesEstimatedWastedLoadedPixels,
+      actualSizesEstimatedWastedLoadedBytes,
+      wastedLoadedPercent
+    };
+  }
+);
+} catch (e) {
+  return [];
+}
+''';
+
+CREATE TEMPORARY FUNCTION IS_CMS(technologies ARRAY<STRUCT<technology STRING, categories ARRAY<STRING>, info ARRAY<STRING>>>, cms STRING, version STRING) RETURNS BOOL AS (
+  EXISTS(
+    SELECT * FROM UNNEST(technologies) AS technology, UNNEST(technology.info) AS info
+    WHERE technology.technology = cms
+    AND (
+      version = ""
+      OR ENDS_WITH(version, ".x") AND (STARTS_WITH(info, RTRIM(version, "x")) OR info = RTRIM(version, ".x"))
+      OR info = version
+    )
+  )
+);
+
+WITH wordpressSizesData AS (
+  SELECT
+    client,
+    image
+  FROM
+    `httparchive.all.pages`,
+    UNNEST(getSrcsetSizesAccuracy(payload)) AS image
+  WHERE
+    date = '2024-02-01'
+    AND IS_CMS(technologies, 'WordPress', '')
+    AND is_root_page = TRUE
+)
+
+SELECT
+  percentile,
+  client,
+  APPROX_QUANTILES(image.sizesAbsoluteError, 100)[OFFSET(percentile)] AS sizesAbsoluteError,
+  APPROX_QUANTILES(image.sizesRelativeError, 100)[OFFSET(percentile)] AS sizesRelativeError,
+  APPROX_QUANTILES(image.wDescriptorAbsoluteError, 100)[OFFSET(percentile)] AS wDescriptorAbsoluteError,
+  APPROX_QUANTILES(image.wDescriptorRelativeError, 100)[OFFSET(percentile)] AS wDescriptorRelativeError,
+  APPROX_QUANTILES(image.actualSizesEstimatedWastedLoadedPixels, 100)[OFFSET(percentile)] AS actualSizesEstimatedWastedLoadedPixels,
+  APPROX_QUANTILES(image.actualSizesEstimatedWastedLoadedBytes, 100)[OFFSET(percentile)] AS actualSizesEstimatedWastedLoadedBytes,
+  APPROX_QUANTILES(image.wastedLoadedPercent, 100)[OFFSET(percentile)] AS wastedLoadedPercent
```

Comment:
It may be helpful to add comments above some of these lines for what the values describe. As far as I understand from reading https://github.com/HTTPArchive/custom-metrics/blob/main/dist/responsive_images.js:
* The first two `sizes*Error` values describe "how wrong" the `sizes` attribute is based on the current client dimensions. There will likely almost always be some degree of mistake here - though that's not necessarily that crucial, since we're limited by the available `srcset` candidates anyway.
* The `wDescriptor*Error` values describe "how wrong" the `srcset` attribute's `w` values for the selected candidate is. While this is interesting data, it's not really relevant to our current effort, since optimizing `srcset` is out of scope and we can assume WordPress is doing at least a decent job in that since the exact image dimensions are known for attachments.
* The remaining 3 columns take into account the available `srcset` candidates, so this is arguably the only meaningful data when it comes to our current effort of improving the `sizes` attribute.

Maybe we should actually reduce columns to only include the data we really care about for this effort (i.e. the last 3 columns). It'll also reduce noise in the query result, as it will already have a lot of rows due to the many percentiles included.